op-supervisor/supervisor/backend/rewinder/rewinder.go

```diff
@@ -0,0 +1,239 @@
+	}
+	if err := r.attemptRewindUnsafe(ev.ChainID, ev.BadBlockHeight); err != nil {
+		return fmt.Errorf("failed to rewind unsafe chain %s: %w", ev.ChainID, err)
+	}
+	return nil
+}
+
+// attemptRewind attempts to rewind the local and cross databases for the given chain and safety level
+func (r *Rewinder) attemptRewind(chainID eth.ChainID, badBlockHeight uint64, ctrl rewindController) error {
+	// First get the finalized head
+	finalizedHead, err := ctrl.getFinalized(chainID)
+	if err != nil {
+		if errors.Is(err, types.ErrFuture) {
+			finalizedHead = types.BlockSeal{}
+		} else {
+			return fmt.Errorf("failed to get finalized head for chain %s: %w", chainID, err)
+		}
+	}
+
+	// If the bad block's parent is before the finalized head then stop
+	parentHeight := badBlockHeight - 1
+	if parentHeight < finalizedHead.Number {
+		r.log.Warn("requested head is not ahead of finalized head", "chain", chainID, "requested", parentHeight, "finalized", finalizedHead.Number)
+		return nil
+	}
+
+	// Find the latest common newHead between the parent and the finalized head
+	newHead, err := r.findLatestCommonAncestor(chainID, parentHeight, finalizedHead)
+	if err != nil {
+		return fmt.Errorf("failed to find common ancestor for chain %s: %w", chainID, err)
+	}
+
+	// Reset the cross db if it's newer than the newHead
+	crossHead, err := ctrl.getCross(chainID)
+	if err != nil {
+		return fmt.Errorf("failed to get cross for chain %s: %w", chainID, err)
+	}
+	if crossHead.Number >= newHead.Number {
+		r.log.Info("rewinding cross", "safety", ctrl.safetyStr(), "chain", chainID, "newHead", newHead.Number)
+		if err := ctrl.rewindCross(chainID, newHead); err != nil {
+			return fmt.Errorf("failed to rewind cross for chain %s: %w", chainID, err)
+		}
+	}
+
+	// Rewind the local db if it's newer than the newHead
+	localHead, err := ctrl.getLocal(chainID)
+	if err != nil {
+		return err
+	}
+	if localHead.Number >= newHead.Number {
+		r.log.Info("rewinding local", "safety", ctrl.safetyStr(), "chain", chainID, "newHead", newHead.Number)
+		if err := ctrl.rewindLocal(chainID, newHead); err != nil {
+			return fmt.Errorf("failed to rewind local for chain %s: %w", chainID, err)
+		}
+	}
+
+	return nil
+}
+
+// attemptRewindUnsafe attempts to rewind unsafe blocks
+func (r *Rewinder) attemptRewindUnsafe(chainID eth.ChainID, badBlockHeight uint64) error {
+	return r.attemptRewind(chainID, badBlockHeight, rewindController{
+		isSafe:       false,
+		getFinalized: derivedFromPairGetter(r.db.LocalSafe),
+		getLocal:     r.db.LocalUnsafe,
+		getCross:     r.db.CrossUnsafe,
+		rewindLocal:  r.db.RewindLocalUnsafe,
+		rewindCross:  r.db.RewindCrossUnsafe,
+	})
+}
+
+// attemptRewindSafe attempts to rewind safe blocks
+func (r *Rewinder) attemptRewindSafe(chainID eth.ChainID, badBlockHeight uint64) error {
+	return r.attemptRewind(chainID, badBlockHeight, rewindController{
+		isSafe:       true,
+		getFinalized: r.db.Finalized,
```

Comment:
Why does RewindUnsafe have a `getFinalized` of `derivedFromPairGetter(r.db.LocalSafe)`? I would assume they have the same Finalized fetching?

Reply:
This is because for Safe reorgs we want to not rewind past the Finalized head, but for Unsafe reorgs we want to not rewind past the LocalSafe head. I've updated the naming to make this more clear by calling it "minBlock" instead of "finalized".